Allow non CS entropy sources for `psa_generate_random()`

valeriosetti opened the pull request:
## Workflow of this PR

- starts with some renaming of entropy kconfigs in order to have them more aligned to Mbed TLS build symbols (which simplifies understanding their meaning/effect).
- improves the description of PSA kconfigs related to RNGs.
- align entropy gathering between legacy `mbedtls_hardware_poll()` and PSA `mbedtls_psa_external_get_random()`.
- and finally it adds a new kconfig `CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CS` to allow non-CS entropy sources also for PSA
    - this is meant to be used only for tests (ex: testing a psa_get_random()` also on platforms that do not have CS sources), but it must be disabled in production. A warning is raised whenever this feature is enabled.

## Improvements from this PR

Prior to this PR when Mbed TLS's PSA crypto was to be built:

- `MBEDTLS_PSA_CRYPTO_LEGACY_RNG` was to be used on platforms that do not have an entropy driver
    - this required also `MBEDTLS_ENTROPY_ENABLED` to add a Zephyr's entropy callback from Mbed TLS's entropy module
- `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` was to be used on platforms that have an entropy driver.

After this PR it will be possible to enable `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` + `CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG` on all platforms. The latter symbols must be disabled in production of course, but for CI purposes it allows a more uniform usage of Mbed TLS's PSA crypto across the Zephyr's tessts and samples.

valeriosetti committed:
mbedtls: rename MBEDTLS_ENTROPY_ENABLED as MBEDTLS_ENTROPY_C

ENABLED suffix does not really makes much sense for a kconfig
so it should be removed. Also other MBEDTLS symbols were recently
updated accordingly.
Moreover having it named exactly the same way as in Mbed TLS
symplifies the understanding of what this kconfig is doing.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
mbedtls: rename CONFIG_MBEDTLS_ZEPHYR_ENTROPY and move it

- Slightly improve the name of this kconfig adding the suffix
  _POLL in order to highlight that this adds a callback
  function used to poll data.

- Description was also updated to point out that this symbols
  might not only use the (secure) entropy driver, but also
  generic number generators, some of which are not really
  secure.

- The symbol was move from Kconfig to Kconfig.tls-generic because
  this is where MBEDTLS_ENTROPY_C is located and since
  MBEDTLS_ENTROPY_HARDWARE_ALT depends on the former (it only
  makes sense if the entropy module is also enabled), we add
  also the "depends on".

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
mbedtls: improve description of PSA random number generators

Improve the description of both MBEDTLS_PSA_CRYPTO_LEGACY_RNG and
MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG kconfig symbols.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
mbedtls: add new kconfig for non-CSPRNG sources in psa_generate_random()

This is meant to be used only for tests on platforms where CSPRNG
sources are not available. It should not be used in production.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
test: mbedtls: new test for PSA crypto initialization and RNG configuration

Add a simple test to showcase how RNG can be configured on different
platforms in order to allow Mbed TLS's PSA crypto implementation to
work properly.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font commented on modules/mbedtls/Kconfig.tls-generic:
```diff
@@ -387,12 +387,22 @@ config MBEDTLS_HAVE_ASM
 	  mostly used in conjunction with CTR_DRBG or HMAC_DRBG to create
 	  a deterministic random number generator.
 
+config MBEDTLS_ZEPHYR_ENTROPY_POLL
```
Thinking about the naming of this... Would `MBEDTLS_POLL_ZEPHYR_ENTROPY` (or `MBEDTLS_ENTROPY_POLL_ZEPHYR` express more clearly that `Mbed TLS polls Zephyr for entropy`?

tomi-font commented on tests/modules/uoscore/prj.conf:
```diff
@@ -12,7 +12,7 @@ CONFIG_UOSCORE=y
 CONFIG_MBEDTLS_ENABLE_HEAP=y
 CONFIG_MBEDTLS_HEAP_SIZE=2048
-CONFIG_MBEDTLS_ZEPHYR_ENTROPY=y
+CONFIG_MBEDTLS_ZEPHYR_ENTROPY_POLL=y
```
```suggestion
CONFIG_MBEDTLS_ENTROPY_C=y
CONFIG_MBEDTLS_ZEPHYR_ENTROPY_POLL=y
```

tomi-font commented on modules/mbedtls/Kconfig.tls-generic:
```diff
@@ -483,16 +483,33 @@ config MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG
+	  requires a cryptographically secure random number generator (CSPRNG)
+	  to be available in the system. If no CSPRNG is available:
+	  - there is no fallback to weak entropy random generators.
+	  - the PSA crypto initialization will fail and none of the PSA APIs
```
Thinking out loud: could we find out about this at build time to avoid failing only at runtime?

tomi-font commented on modules/mbedtls/Kconfig.tls-generic:
```diff
@@ -467,21 +481,49 @@ choice MBEDTLS_PSA_CRYPTO_RND_SOURCE
+	  requires a cryptographically secure random number generator (CSPRNG)
+	  to be available in the system. If no CSPRNG is available:
+	  - there is no fallback to weak entropy random generators.
+	  - the PSA crypto initialization will fail and none of the PSA APIs
+	    will not be available.
```
```suggestion
	  - the initialization of PSA Crypto will fail and none of its API
	    will be available.
```

tomi-font commented on modules/mbedtls/Kconfig.tls-generic:
```diff
@@ -467,21 +481,49 @@ choice MBEDTLS_PSA_CRYPTO_RND_SOURCE
+	  Security level in this case really depends on the type of entropy
+	  sources which are enabled in the system: if weak entropy sources are
+	  used, then the generated data will only be pseudo random. Strong
+	  entropy sources are really suggested (if possible) to have real
```
```suggestion
	  entropy sources are strongly recommended (if possible) to have real
```

tomi-font commented on modules/mbedtls/Kconfig.tls-generic:
```diff
@@ -467,21 +481,49 @@ choice MBEDTLS_PSA_CRYPTO_RND_SOURCE
+	  used, then the generated data will only be pseudo random. Strong
+	  entropy sources are really suggested (if possible) to have real
+	  random data.
+	  The main difference between this implementation and the
+	  MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG counterpart is just in terms of
+	  code size: this brings in more legacy modules and therefore it has
```
Really? Aren't there other significant differences? (Some of them are stated in you new description of `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG`.)

valeriosetti replied:
Right, the initial part of the sentence ("the main difference") moves the focus away from security. I rephrased the sentence.

tomi-font commented on modules/mbedtls/CMakeLists.txt:
```diff
@@ -11,6 +11,18 @@ zephyr_interface_library_named(mbedTLS)
+        Warning: non cryptographycally secure sources are enabled for psa_generate_random().
+        This is meant to be used only for tests, not in production!")
+    else()
+      if(CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG AND NOT CONFIG_CSPRNG_ENABLED)
+        message(WARNING "
+          MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG is set but there is
+          no CSPRNG enabled.")
+      endif()
```
Couldn't we just make the build fail in this case? If `EXTERNAL_RNG` is enabled, CSPRNG is not enabled, nor is `RNG_ALLOW_NON_CSPRNG`, then `psa_crypto_init()` will always fail right?

tomi-font commented on modules/mbedtls/CMakeLists.txt:
```diff
@@ -11,6 +11,18 @@ zephyr_interface_library_named(mbedTLS)
 
+    if(CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG)
+      message(WARNING "
+        Warning: non cryptographycally secure sources are enabled for psa_generate_random().
```
I think no need for `Warning:` here because CMake will already indicate that.

valeriosetti replied:
Yeah, but I saw [somewhere else](https://github.com/zephyrproject-rtos/zephyr/blob/c008cbab1a05316139de191b0553ab6ccc0073ad/subsys/random/CMakeLists.txt#L12) where the "Warning:" part was still added to the message so I thought I could do the same. However I don't have any strong opinion on this, so I can remove that ;)

tomi-font commented on modules/mbedtls/Kconfig.tls-generic:
```diff
@@ -513,6 +512,18 @@ config MBEDTLS_PSA_CRYPTO_LEGACY_RNG
 
 endchoice
 
+config MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG
```
Not sure about this suggestion, but this new Kconfig option could potentially be added as part of the `choice`, removing the dependency with `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` (and putting put the dependency of `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` on `CSPRNG_ENABLED`)

valeriosetti replied:
But this would make `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG` as the 3rd option of the choice, whereas my purpose here is to make it as a "feature" of `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG`.

tomi-font commented on modules/mbedtls/zephyr_init.c:
Not introduced here, but this file has more than just initialization... Could rename/split it?

tomi-font commented on modules/mbedtls/zephyr_init.c:
```diff
@@ -134,13 +141,17 @@ psa_status_t mbedtls_psa_external_get_random(
 
-	if (sys_csrand_get(output, output_size) != 0) {
+	ret = get_random_data(output, output_size,
+		IS_ENABLED(CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG) ?
+		false : true);
```
Suggest simplifying this by removing the ternary operator and turning `cs_only` in `get_random_data()` into `allow_non_cs`:
```suggestion
		IS_ENABLED(CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG));
```

tomi-font commented on modules/mbedtls/zephyr_init.c:
```diff
@@ -122,6 +84,51 @@ mbedtls_ms_time_t mbedtls_ms_time(void)
+#if defined(CONFIG_MBEDTLS_ZEPHYR_ENTROPY_POLL) || defined(CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
+static int get_random_data(uint8_t *output, size_t output_size, bool cs_only)
+{
+	int ret = MBEDTLS_ERR_ENTROPY_SOURCE_FAILED;
```
```suggestion
	int ret = MBEDTLS_ERR_ENTROPY_NO_SOURCES_DEFINED;
```

tomi-font commented on modules/mbedtls/zephyr_init.c:
```diff
@@ -122,6 +84,51 @@ mbedtls_ms_time_t mbedtls_ms_time(void)
+		return MBEDTLS_ERR_ENTROPY_SOURCE_FAILED;
+	}
+
+	ret = get_random_data(output, len, false);
```
Noting that here (in `mbedtls_hardware_poll()`) we're completely dropping the use of `CONFIG_ENTROPY_HAS_DRIVER`/`entropy_get_entropy()` (replacing it by `CONFIG_CSPRNG_ENABLED`/`sys_csrand_get()`).
Is this intended and if so why?

valeriosetti replied:
As for what it concerns with the kconfigs, `CSPRNG_ENABLED` is set whenever `ENTROPY_HAS_DRIVER` [is set](https://github.com/zephyrproject-rtos/zephyr/blob/95622e49b215bcd3ff61f666db6d74eedd4f5b51/subsys/random/Kconfig#L83) and there is no way to disable the `random` subsys so IMO they can be considered as equivalent.

I think that the behavior of `mbedtls_hardware_poll()` is not changed by the new implementation. In the old design the entropy device was polled first (if available) and then `sys_rand_get()` was used as backup. In the new fashion I try `sys_rand_get()` first (again, if available) and then I use `sys_rand_get()` as fallback.
In facts `sys_rand_get()` is implemented by:
- `zephyr/subsys/random/random_ctr_drbg.c` and
- `zephyr/subsys/random/random_entropy_device.c`.
Both implementations get the data from an entropy driver.

In conclusion I think that my implementation basically re-implements the same pattern that we had previously. Am I wrong?

tomi-font replied:
I don't think so. I was just checking with you the logic. All good!

tomi-font commented on tests/crypto/mbedtls_psa/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(mbedtls)
+
+FILE(GLOB app_sources src/*.c)
+target_sources(app PRIVATE ${app_sources})
```
Not sure globbing is allowed. You may as well replace this by just
```suggestion
target_sources(app PRIVATE src/main.c)
```

tomi-font commented on tests/crypto/mbedtls_psa/testcase.yaml:
```diff
@@ -0,0 +1,37 @@
+# The goal here is to showcase that Mbed TLS's PSA implementation can be
```
I think it could help to make sure that we talk about things naming them properly, i.e. here PSA Crypto instead of just PSA, in order to avoid spreading more confusion to the world.
```suggestion
# The goal here is to showcase that Mbed TLS's PSA Crypto implementation can be
```
And I have the same suggestion for this commit's message. Too bad we can't make suggestions on commit messages.
(Actually, there are many occurences of PSA below that could be changed to PSA Crypto. Won't list them individually.)

tomi-font commented on tests/crypto/mbedtls_psa/testcase.yaml:
```diff
@@ -0,0 +1,37 @@
+# - psa_crypto_init() which is required before any PSA operation
+# - psa_generate_random() which is always available as long as PSA is initialized.
+#
+# Since it might take too long to really execute this test on all Zephyr's
+# supported platforms, we reduce to a very small selected group whose
```
```suggestion
# Since it might take too long to execute this test on all platforms
# supported by Zephyr, we reduce to a very small selected group whose
```

tomi-font commented on tests/crypto/mbedtls_psa/testcase.yaml:
```diff
@@ -0,0 +1,37 @@
+tests:
+  crypto.mbedtls_psa.with_entropy_driver:
+    # Pick a platform which has an entropy driver and enable it.
+    platform_allow:
```
I think that instead of using `platform_allow` you could use `integration_platforms`. Especially since the behavior of `platform_allow` may be surprising in certain cases (https://github.com/zephyrproject-rtos/zephyr/pull/75962). See https://docs.zephyrproject.org/latest/develop/test/twister.html for reference.

valeriosetti replied:
Thanks for the tip! I didn't know the difference between `platform_allow` and `integration_platform` before.

tomi-font commented on tests/crypto/mbedtls_psa/src/main.c:
```diff
@@ -0,0 +1,26 @@
+
+#include <zephyr/ztest.h>
+
+#include "psa/crypto.h"
```
```suggestion
#include <psa/crypto.h>
```

tomi-font reviewed (commented):
Good stuff!

valeriosetti committed:
mbedtls: move entropy polling functions to a dedicated file

This commit just moves some code around. For sake of cleanliness
a new file is added to hold the entropy parsing functions
for Mbed TLS.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font commented:
@valeriosetti Actually, I didn't think of this earlier, but this probably deserves some migration guide entry at least for the renamed Kconfig options.
